packages/itwinui-react/src/core/Table/Table.tsx

```diff
@@ -578,18 +580,44 @@ export const Table = <
     );
   }, [data, getSubRows]);
 
-  const [rowHasParent] = React.useState(() => new WeakSet<T>());
-
   const getSubRowsWithSubComponents = React.useCallback(
-    (originalRow: T) => {
-      if (!rowHasParent.has(originalRow)) {
-        rowHasParent.add(originalRow);
-        return [originalRow];
+    (originalRow: T, relativeIndex: number) => {
+      // If originalRow represents a subcomponent, don't add any subrows to it
+      if (originalRow[iuiId as keyof typeof originalRow]) {
+        return [];
       }
 
-      return (originalRow.subRows as T[]) || [];
+      // If originalRow represents the top-most level row, add itself as a subrow that will represent a subComponent.
+      // Add a symbol as a key on this subrow to distinguish it from the real row.
+      // This distinguishment is needed as the subrow needs to have all fields of the row since react-table expects even subrows to be rows (Row<T>).
+      return (
+        (originalRow.subRows as T[]) ?? [
+          {
+            [iuiId]: `subcomponent-${relativeIndex}`,
+            ...originalRow,
+          },
```

Comment:
Spreading can be expensive/slow. I think we could do this instead:

```js
if (originalRow.subRows) {
  return originalRow.subRows;
}

originalRow[iuiId] = `subcomponent-${relativeIndex}`;
return originalRow;
```

Reply:
![image](https://github.com/user-attachments/assets/ab9c18a5-4336-49e6-90f7-bd1183dd5e23)

Hmm, I got this error while trying to assign the value. Is there any other way to approach this?

Reply:
This is just a TS issue, which we can ignore (assuming the JS code still works correctly).

```js
(originalRow as any)[iuiId] = `subcomponent-${relativeIndex}`;
```

Reply:
![image](https://github.com/user-attachments/assets/d5a17c81-68af-499c-a0e5-d3cef75f382b)

I run into this issue where it creates the right sub-rows in the first render but in the next one, since the original rows has the Symbol key and each symbol is unique, it keeps adding the symbols into the original data. When it loops back and hits the first if statement, it returns an empty sub-row array.

Reply:
That doesn't look right. Each symbol is _not_ unique; it's the same symbol.

![](https://github.com/user-attachments/assets/3f586f33-82f8-477b-ab9e-750ad0743f55)

Did you test this change *after* moving the symbol out of the component (based on the [other comment](https://github.com/iTwin/iTwinUI/pull/2306#discussion_r1806552426))?

Reply:
Right, I made this commit after moving the Symbol out commit.

Reply:
I looked into this and realized that my suggestion was flawed. We don't want to modify the original row, we want to return a copy of it, because the expanded-content is technically a new row.

So let's go with the original solution:

```js
if (originalRow.subRows) {
  return originalRow.subRows as T[];
}

return [
  {
    [iuiId]: `subcomponent-${relativeIndex}`,
    ...originalRow,
  },
];
```

Reply:
I have reverted back to the original solution.